website/blog/2023-08-15-how-opendal-read-data/index.md

```diff
@@ -0,0 +1,158 @@
+
+- Free from services: Any service can be accessed freely through native interfaces
+- Free from implementations: No matter how the underlying implementation is, it can be called in a unified way
+- Free to integrate: Able to freely integrate with various services and languages
+- Free to zero cost: Users don't have to pay for features they don't use
+
+On this philosophical foundation, OpenDAL Rust Core can be mainly divided into the following components:
+
+- Operator: The outer interface exposed to users
+- Layers: Specific implementation of different middleware
+- Services: Specific implementation of different services
+
+From a macroscopic perspective, OpenDAL's data reading call stack would look like this:
+
+![](1.png)
+
+All Layers and Services have implemented a unified Accessor interface, erasing all type information when building the Operator. For the Operator, regardless of what services are used or how many middleware are added, all call logic is consistent. This design splits OpenDAL's API into Public API and Raw API, where the Public API is directly exposed to users, providing convenient top-level interfaces, and Raw API is provided to OpenDAL internal developers, maintaining a unified internal interface and providing some convenient implementation.
+
+## Operator
+
+OpenDAL's Operator API will adhere to a consistent calling paradigm as much as possible, reducing users' learning and usage costs. For example, OpenDAL offers the following APIs for `read`:
+
+- `op.read(path)`: Reads the entire content of the specified file
+- `op.reader(path)`: Creates a Reader for streaming reading
+- `op.read_with(path).range(1..1024)`: Reads file content using specified parameters, such as range
+- `op.reader_with(path).range(1..1024)`: Creates a Reader for streaming reading with specified parameters
+
+It's not hard to see that `read` is more like syntactic sugar, allowing users to quickly read files without considering various traits like `AsyncRead`. The `reader` provides more flexibility, implementing widely-used community traits like `AsyncSeek`, `AsyncRead`, allowing more flexible data reading. `read_with` and `reader_with` assist users in specifying various parameters in a more natural way through Future Builder functions.
+
+The internal logic of the Operator would look like this:
+
+![](2.png)
+
+Its main job is to encapsulate the interface for the user:
+
+- Completing the construction of `OpRead`
+- Calling the `read` function provided by `Accessor`
+- Wrapping the returned value as `Reader` and implementing interfaces like `AsyncSeek`, `AsyncRead`, etc., based on `Reader`
+
+## Layers
+
+A little secret here is that OpenDAL will automatically apply some Layers to the Service to implement some internal logic. As of the completion of this article, OpenDAL's automatically added Layers include:
+
+- `ErrorContextLayer`: Injects context information, such as `scheme`, `path`, etc., into all returned errors of Operation
```

Comment:
What is the scheme?

Reply:
This article is not intended for users who have no knowledge about OpenDAL...